sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -75,24 +87,46 @@ internal class OperationInternal : OperationInternalBase
             string? operationTypeName = null,
             IEnumerable<KeyValuePair<string, string>>? scopeAttributes = null,
             DelayStrategy? fallbackStrategy = null)
-            :base(clientDiagnostics, rawResponse, operationTypeName ?? operation.GetType().Name, scopeAttributes, fallbackStrategy)
+            :base(clientDiagnostics, operationTypeName ?? operation.GetType().Name, scopeAttributes, fallbackStrategy)
         {
-            _operation = operation;
+            _internalOperation = new OperationInternal<VoidValue>(clientDiagnostics, new OperationWrapper(operation), rawResponse, operationTypeName ?? operation.GetType().Name, scopeAttributes, fallbackStrategy);
         }
 
-        /// <summary>
-        /// Sets the <see cref="OperationInternal"/> state immediately.
-        /// </summary>
-        /// <param name="state">The <see cref="OperationState"/> used to set <see cref="OperationInternalBase.HasCompleted"/> and other members.</param>
-        public void SetState(OperationState state)
+        private OperationInternal(OperationState finalState)
+            :base(finalState.RawResponse)
         {
-            ApplyStateAsync(false, state.RawResponse, state.HasCompleted, state.HasSucceeded, state.OperationFailedException, throwIfFailed: false).EnsureCompleted();
+            _internalOperation = finalState.HasSucceeded
+                ? OperationInternal<VoidValue>.Succeeded(finalState.RawResponse, default)
+                : OperationInternal<VoidValue>.Failed(finalState.RawResponse, finalState.OperationFailedException!);
         }
 
-        protected override async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        public override Response RawResponse => _internalOperation.RawResponse;
+
+        public override bool HasCompleted => _internalOperation.HasCompleted;
+
+        protected override async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken) =>
+            async ? await _internalOperation.UpdateStatusAsync(cancellationToken).ConfigureAwait(false) : _internalOperation.UpdateStatus(cancellationToken);
+
+        private readonly struct VoidValue { }
+
+        private readonly struct OperationWrapper : IOperation<VoidValue>
```

Comment:
Shoudl this be a class?

Reply:
Let's give this a name that implies it converts it to a generic, and/or have a good comment here -- wrapper could mean lots of things.